### DISCUSSION TITLE
How to read large files in nodejs (v16)

### DISCUSSION THREAD

**Discussion by gowy222** [gildas-lormeau/zip.js#420] [Q&A]
const zipReader = new zip.ZipReader(readable)

ERR:
reader = new BlobReader(await new Response(reader.readable).blob()); //put all data into memory? if file over 2GB..
=> ReferenceError: Response is not defined

----
Response  is Web API? is there any  way work with big file(>2GB)  use fs.createReadStream(local_path)... low memory! ; 
and noted NO Readable.toWeb() in pure nodejs tho..

**Comment by gildas-lormeau** [comment]
If you cannot use [web streams](https://nodejs.org/api/webstreams.html) natively, you can write a class extending the [`Reader`](https://gildas-lormeau.github.io/zip.js/api/classes/Reader.html) class of zip.js and use it to read data. 

You have 2 methods to implement (you can omit the constructor):
- The method `init` should initialize `this.size` with the size in bytes returned by `filehandle#stat` for example, see https://nodejs.org/api/fs.html#filehandlestatoptions.
- The method `readUint8Array` should rely on `fs.createReadStream` and use the options `start`, `end` in order to read sections of the file as `Uint8Array` instances, see https://nodejs.org/api/all.html#fs_fs_createreadstream_path_options. 

There is also an example of such a custom `Reader` class in the documentation or below. 

https://github.com/gildas-lormeau/zip.js/blob/151b5984728dddad42f2f87da39b1efc9b9b7519/tests/all/test-custom-io.js#L6-L25

**Comment by TKdevlop** [comment]
Hi sorry, I am new to this lib and still not sure how to fix the above issue
for example to fix ReferenceError: TransformStream is not defined

need to add this

```
import { TransformStream } from "node:stream/web";

//@ts-ignore
globalThis.TransformStream = TransformStream;
```


can you suggest a fix for the above would be greatly appreciated?

**Reply by gildas-lormeau** [reply]
Didn't the code you added solve the problem? The alternative is to update Node.js.
